logrus.go

```diff
@@ -0,0 +1,131 @@
+package logformatter
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/genproto/googleapis/devtools/clouderrorreporting/v1beta1"
+	"google.golang.org/genproto/googleapis/logging/type"
+)
+
+type (
+	// Interface for inspecting error objects recursively
+	// Ref: https://github.com/pkg/errors/blob/7f95ac13edff643b8ce5398b6ccab125f8a20c1a/errors.go#L30-L53
+	causer interface {
+		Cause() error
+	}
+
+	// Interface for retrieving stack frame for each error object
+	// Ref: https://github.com/pkg/errors/blob/7f95ac13edff643b8ce5398b6ccab125f8a20c1a/errors.go#L68-L73
+	stackTracer interface {
+		StackTrace() errors.StackTrace
+	}
+)
+
+// Adapt from https://github.com/googleapis/google-cloud-go/issues/1084#issuecomment-474565019
+func FormatStack(err error) (buffer []byte) {
+	if err == nil {
+		return nil
+	}
+
+	// find the inner most error with a stack
+	inner := err
+	for inner != nil {
+		if cause, ok := inner.(causer); ok {
+			inner = cause.Cause()
+			if _, ok := inner.(stackTracer); ok {
+				err = inner
+			}
+		} else {
+			break
+		}
+	}
+
+	if stackTrace, ok := err.(stackTracer); ok {
+		buf := bytes.Buffer{}
+		buf.WriteString(getGoroutineState() + "\n")
+
+		// format each frame of the stack to match runtime.Stack's format
+		var lines []string
+		for _, frame := range stackTrace.StackTrace() {
+			pc := uintptr(frame) - 1
+			fn := runtime.FuncForPC(pc)
+			if fn != nil {
+				file, line := fn.FileLine(pc)
+				lines = append(lines, fmt.Sprintf("%s()\n\t%s:%d +%#x", fn.Name(), file, line, fn.Entry()))
+			}
+		}
+		buf.WriteString(strings.Join(lines, "\n"))
+
+		buffer = buf.Bytes()
+	}
+	return
+}
+
+func NewStackdriverFormatter(service, version string) *Stackdriver {
```

Comment:
Could we have some comments for exported functions?
Maybe provide some use cases here?

Reply:
Please see the example_test.go This is the convention how the go showing examples.
https://blog.golang.org/examples.

Reply:
In example_test.go, 
It's a little bit hard to know what `example` and `test` mean.

Reply:
Actually, the ExampleXXX() will provide the example under function XXX() if the godoc is built. This file is indeed tested during `go test` but it always pass as I did not define the output. As I said, this the convention of go. You can see this conventions spread in the standard libraries.

Reply:
I know, just want mention that `example` and `test` are not clear for me to read the doc.